src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+use crate::config::{AllowedFns, Configuration};
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+use futures::future::BoxFuture;
+use holochain_client::{
+    AdminWebsocket, AppWebsocket, AuthorizeSigningCredentialsPayload, CellInfo, ClientAgentSigner,
+    ConductorApiError, ConnectRequest, GrantedFunctions, IssueAppAuthenticationTokenPayload,
+    Timestamp, WebsocketConfig,
+};
+use holochain_types::app::InstalledAppId;
+use holochain_types::websocket::AllowedOrigins;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, RwLock};
+
+/// The origin that the gateway will use when connecting to Holochain app interfaces.
+pub const HTTP_GW_ORIGIN: &str = "hc-http-gw";
+
+/// A wrapper around an app websocket connection that includes state required to manage the
+/// connection.
+#[derive(Debug, Clone)]
+pub struct AppWebsocketWithState {
+    /// The app websocket connection.
+    pub app_ws: AppWebsocket,
+    /// The time at which the connection was opened.
+    pub opened_at: Timestamp,
+}
+
+/// A connection pool for app connections.
+///
+/// This is a pool in the sense that it manages multiple connections to Holochain app interfaces,
+/// but it will manage exactly one connection per installed app.
+#[derive(Debug, Clone)]
+pub struct AppConnPool {
+    configuration: Configuration,
+    cached_app_port: Arc<RwLock<Option<u16>>>,
+    app_clients: Arc<tokio::sync::RwLock<HashMap<InstalledAppId, AppWebsocketWithState>>>,
+}
+
+impl AppConnPool {
+    /// Create a new app connection pool with the given configuration.
+    pub fn new(configuration: Configuration) -> Self {
+        Self {
+            configuration,
+            cached_app_port: Default::default(),
+            app_clients: Default::default(),
+        }
+    }
+
+    /// Call a function with an app client for the given installed app ID.
+    ///
+    /// This function takes care of reconnecting to the app client if the connection is lost. Your
+    /// function is free to operate on the app client without worrying about the connection state.
+    ///
+    /// Exactly one attempt is made to reconnect and re-run the provided callback. If there is a
+    /// second failure after reconnecting, the error is returned to the caller.
+    pub async fn call<T>(
+        &self,
+        installed_app_id: InstalledAppId,
+        admin_websocket: AdminWebsocket,
```

Comment:
Once the admin websocket PR is merged, perhaps it would make sense to add admin ws management to this pool instead of injecting it?

Reply:
Agreed, this needs updating to integrate with the admin websocket PR